File: Khandoba Secure Docs/Services/SupabaseService.swift
//
//  SupabaseService.swift
//  Khandoba Secure Docs
//
//  Created for Supabase Migration
//

import Foundation
import Combine
import Supabase

@MainActor
final class SupabaseService: ObservableObject {
    @Published var isConnected = false
    @Published var currentSession: Session?
    @Published var error: Error?
    
    private var supabaseClient: SupabaseClient?
    private var realtimeChannels: [RealtimeChannel] = []
    
    nonisolated init() {}
    
    // MARK: - Initialization
    
    func configure() async throws {
        let config = SupabaseConfig.currentEnvironment
        
        // Configure AuthClient to use new session behavior (suppresses warning)
        // The default storage (UserDefaults) will be used automatically
        let client = SupabaseClient(
            supabaseURL: URL(string: config.supabaseURL)!,
            supabaseKey: config.supabaseAnonKey,
            options: SupabaseClientOptions(
                auth: .init(
                    emitLocalSessionAsInitialSession: true
                )
            )
        )
        
        self.supabaseClient = client
        
        // Check connection by testing a simple database query (doesn't require auth)
        do {
            // Test connection with a simple query that doesn't require authentication
            // This verifies the Supabase URL and key are correct
            let _: [SupabaseUser] = try await client.database
                .from("users")
                .select()
                .limit(0)
                .execute()
                .value
            
            self.isConnected = true
            self.currentSession = try? await client.auth.session
            print("✅ Supabase client initialized and connected")
            print("   URL: \(config.supabaseURL)")
            print("   Connection verified via database query")
        } catch {
            // If query fails, check if it's an auth error (expected) or connection error
            if let urlError = error as? URLError {
                self.isConnected = false
                self.error = error
                print("❌ Supabase connection failed: \(urlError.localizedDescription)")
                print("   Check your internet connection and Supabase URL")
            } else {
                // Auth errors are expected when not signed in - connection is still OK
                self.isConnected = true
                self.currentSession = try? await client.auth.session
                print("✅ Supabase client initialized")
                print("   URL: \(config.supabaseURL)")
                print("   Note: No active session (user not signed in)")
            }
        }
        
        // Setup real-time subscriptions if enabled and user is authenticated
        // Note: Realtime subscriptions require authentication, so we'll set them up after sign-in
        // Don't set up here to avoid WebSocket errors when not authenticated
    }
    
    // MARK: - Authentication
    
    func signInWithApple(idToken: String, nonce: String) async throws -> Session {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        do {
            let session = try await client.auth.signInWithIdToken(
                credentials: OpenIDConnectCredentials(
                    provider: .apple,
                    idToken: idToken,
                    nonce: nonce
                )
            )
            
            self.currentSession = session
            self.isConnected = true
            
            // Setup real-time subscriptions after successful authentication
            if SupabaseConfig.enableRealtime {
                setupRealtimeSubscriptions()
            }
            
            return session
        } catch {
            self.error = error
            throw error
        }
    }
    
    func signOut() async throws {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        try await client.auth.signOut()
        self.currentSession = nil
        self.isConnected = false
        
        // Unsubscribe from realtime when signing out
        unsubscribeAll()
    }
    
    func getCurrentUser() async throws -> Supabase.User? {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        return try await client.auth.user()
    }
    
    // MARK: - Database Queries
    
    func query(_ table: String) -> PostgrestQueryBuilder {
        guard let client = supabaseClient else {
            fatalError("Supabase client not initialized. Call configure() first.")
        }
        
        return client.database.from(table)
    }
    
    func insert<T: Codable>(_ table: String, values: T) async throws -> T {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        let response: T = try await client.database
            .from(table)
            .insert(values)
            .select()
            .single()
            .execute()
            .value
        
        return response
    }
    
    func update<T: Codable>(_ table: String, id: UUID, values: T) async throws -> T {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        let response: T = try await client.database
            .from(table)
            .update(values)
            .eq("id", value: id.uuidString)
            .select()
            .single()
            .execute()
            .value
        
        return response
    }
    
    func delete(_ table: String, id: UUID) async throws {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        try await client.database
            .from(table)
            .delete()
            .eq("id", value: id.uuidString)
            .execute()
    }
    
    func fetch<T: Codable>(_ table: String, id: UUID) async throws -> T {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        let response: T = try await client.database
            .from(table)
            .select()
            .eq("id", value: id.uuidString)
            .single()
            .execute()
            .value
        
        return response
    }
    
    func fetchAll<T: Codable>(
        _ table: String,
        filters: [String: Any]? = nil,
        limit: Int? = nil,
        orderBy: String? = nil,
        ascending: Bool = true
    ) async throws -> [T] {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        // Build query with filters
        var filterQuery = client.database.from(table).select()
        
        // Apply filters
        if let filters = filters {
            for (key, value) in filters {
                // Convert value to PostgrestFilterValue-compatible type
                if let stringValue = value as? String {
                    filterQuery = filterQuery.eq(key, value: stringValue)
                } else if let uuidValue = value as? UUID {
                    filterQuery = filterQuery.eq(key, value: uuidValue.uuidString)
                } else if let intValue = value as? Int {
                    filterQuery = filterQuery.eq(key, value: intValue)
                } else if let boolValue = value as? Bool {
                    filterQuery = filterQuery.eq(key, value: boolValue)
                } else {
                    // Fallback: convert to string
                    filterQuery = filterQuery.eq(key, value: String(describing: value))
                }
            }
        }
        
        // Build final query with ordering and limit
        // Note: order() and limit() return PostgrestTransformBuilder
        // We need to apply order() first to get a transform builder, then we can apply limit()
        let response: [T]
        if let orderBy = orderBy {
            // Apply order first to get PostgrestTransformBuilder
            let transformQuery = filterQuery.order(orderBy, ascending: ascending)
            if let limit = limit {
                // Both order and limit
                response = try await transformQuery.limit(limit).execute().value
            } else {
                // Only order
                response = try await transformQuery.execute().value
            }
        } else if let limit = limit {
            // Only limit - apply a default order first to get transform builder
            // Using a common column that should exist in most tables
            let transformQuery = filterQuery.order("created_at", ascending: false)
            response = try await transformQuery.limit(limit).execute().value
        } else {
            // No order or limit - execute filter query directly
            response = try await filterQuery.execute().value
        }
        return response
    }
    
    // MARK: - Storage
    
    func uploadFile(
        bucket: String,
        path: String,
        data: Data,
        options: FileOptions? = nil
    ) async throws -> String {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        // Use default options if none provided
        let uploadOptions = options ?? FileOptions(
            cacheControl: "3600",
            contentType: "application/octet-stream",
            upsert: false
        )
        
        try await client.storage.from(bucket).upload(
            path: path,
            file: data,
            options: uploadOptions
        )
        
        return path
    }
    
    func downloadFile(bucket: String, path: String) async throws -> Data {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        let data = try await client.storage.from(bucket).download(path: path)
        return data
    }
    
    func deleteFile(bucket: String, path: String) async throws {
        guard let client = supabaseClient else {
            throw SupabaseError.clientNotInitialized
        }
        
        try await client.storage.from(bucket).remove(paths: [path])
    }
    
    // MARK: - Real-time
    
    private func setupRealtimeSubscriptions() {
        guard let client = supabaseClient else {
            print("⚠️ Cannot setup realtime: Supabase client not initialized")
            return
        }
        
        // Only setup realtime if user is authenticated
        guard currentSession != nil else {
            print("⚠️ Cannot setup realtime: User not authenticated")
            return
        }
        
        Task {
            do {
                for channelName in SupabaseConfig.realtimeChannels {
                    let channel = client.realtime.channel("\(channelName)-changes")
                    realtimeChannels.append(channel)
                    
                    // Subscribe to INSERT events
                    channel.on("postgres_changes", filter: ChannelFilter(
                        event: "INSERT",
                        schema: "public",
                        table: channelName
                    )) { [weak self] message in
                        Task { @MainActor in
                            print("📡 Real-time INSERT on \(channelName)")
                            NotificationCenter.default.post(
                                name: .supabaseRealtimeUpdate,
                                object: nil,
                                userInfo: [
                                    "channel": channelName,
                                    "event": "INSERT",
                                    "payload": message.payload
                                ]
                            )
                        }
                    }
                    
                    // Subscribe to UPDATE events
                    channel.on("postgres_changes", filter: ChannelFilter(
                        event: "UPDATE",
                        schema: "public",
                        table: channelName
                    )) { [weak self] message in
                        Task { @MainActor in
                            print("📡 Real-time UPDATE on \(channelName)")
                            NotificationCenter.default.post(
                                name: .supabaseRealtimeUpdate,
                                object: nil,
                                userInfo: [
                                    "channel": channelName,
                                    "event": "UPDATE",
                                    "payload": message.payload
                                ]
                            )
                        }
                    }
                    
                    // Subscribe to DELETE events
                    channel.on("postgres_changes", filter: ChannelFilter(
                        event: "DELETE",
                        schema: "public",
                        table: channelName
                    )) { [weak self] message in
                        Task { @MainActor in
                            print("📡 Real-time DELETE on \(channelName)")
                            NotificationCenter.default.post(
                                name: .supabaseRealtimeUpdate,
                                object: nil,
                                userInfo: [
                                    "channel": channelName,
                                    "event": "DELETE",
                                    "payload": message.payload
                                ]
                            )
                        }
                    }
                    
                    // Subscribe to the channel with error handling
                    do {
                        await channel.subscribe()
                        print("✅ Subscribed to realtime channel: \(channelName)")
                    } catch {
                        print("⚠️ Failed to subscribe to channel \(channelName): \(error.localizedDescription)")
                        // Continue with other channels even if one fails
                    }
                }
                
                await MainActor.run {
                    print("✅ Real-time subscriptions setup for \(SupabaseConfig.realtimeChannels.count) channels")
                }
            } catch {
                await MainActor.run {
                    print("⚠️ Error setting up realtime subscriptions: \(error.localizedDescription)")
                    print("   This is normal if Realtime isn't enabled in Supabase or tables aren't in publication")
                }
            }
        }
    }
    
    func unsubscribeAll() {
        for channel in realtimeChannels {
            channel.unsubscribe()
        }
        realtimeChannels.removeAll()
    }
}

// MARK: - Errors

enum SupabaseError: LocalizedError {
    case clientNotInitialized
    case invalidResponse
    case networkError(Error)
    case authenticationFailed
    case storageError(String)
    
    var errorDescription: String? {
        switch self {
        case .clientNotInitialized:
            return "Supabase client not initialized. Call configure() first."
        case .invalidResponse:
            return "Invalid response from Supabase"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .authenticationFailed:
            return "Authentication failed"
        case .storageError(let message):
            return "Storage error: \(message)"
        }
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let supabaseRealtimeUpdate = Notification.Name("supabaseRealtimeUpdate")
}

File: platforms/android/app/src/main/java/com/khandoba/securedocs/data/supabase/SupabaseService.kt
package com.khandoba.securedocs.data.supabase

import android.util.Log
import com.khandoba.securedocs.config.AppConfig
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.realtime.realtime
import io.github.jan.supabase.storage.storage
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.auth.providers.oauth.Google
import io.github.jan.supabase.auth.user.UserInfo
import io.github.jan.supabase.auth.session.Session
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class SupabaseUser(
    val id: UUID,
    val google_user_id: String, // Android equivalent of apple_user_id
    val full_name: String,
    val email: String? = null,
    val profile_picture_url: String? = null,
    val created_at: String, // ISO 8601 date string
    val last_active_at: String,
    val is_active: Boolean = true,
    val is_premium_subscriber: Boolean = false,
    val subscription_expiry_date: String? = null,
    val updated_at: String
)

@Serializable
data class SupabaseVault(
    val id: UUID,
    val name: String,
    val vault_description: String? = null,
    val owner_id: UUID,
    val created_at: String,
    val last_accessed_at: String? = null,
    val status: String = "locked",
    val key_type: String = "single",
    val vault_type: String = "both",
    val is_system_vault: Boolean = false,
    val encryption_key_data: String? = null, // Base64 encoded
    val is_encrypted: Boolean = true,
    val is_zero_knowledge: Boolean = true,
    val relationship_officer_id: UUID? = null,
    val is_anti_vault: Boolean = false,
    val monitored_vault_id: UUID? = null,
    val anti_vault_id: UUID? = null,
    val threat_index: Double = 0.0,
    val threat_level: String = "low", // "low", "medium", "high", "critical"
    val last_threat_assessment_at: String? = null,
    val updated_at: String
)

@Serializable
data class SupabaseVaultTransferRequest(
    val id: String,
    val vault_id: String,
    val requested_by_user_id: String,
    val requested_at: String,
    val status: String,
    val reason: String? = null,
    val new_owner_id: String? = null,
    val new_owner_name: String? = null,
    val new_owner_phone: String? = null,
    val new_owner_email: String? = null,
    val transfer_token: String,
    val approved_at: String? = null,
    val approver_id: String? = null,
    val ml_score: Double? = null,
    val ml_recommendation: String? = null,
    val threat_index: Double? = null,
    val completed_at: String? = null,
    val created_at: String,
    val updated_at: String
)

@Serializable
data class SupabaseAntiVault(
    val id: UUID,
    val vault_id: UUID,
    val monitored_vault_id: UUID,
    val owner_id: UUID,
    val status: String = "locked", // "locked", "active", "archived"
    val auto_unlock_policy: Map<String, kotlinx.serialization.json.JsonElement>? = null, // JSONB
    val threat_detection_settings: Map<String, kotlinx.serialization.json.JsonElement>? = null, // JSONB
    val last_intel_report_id: UUID? = null,
    val created_at: String,
    val updated_at: String,
    val last_unlocked_at: String? = null
)

@Serializable
data class SupabaseDocument(
    val id: UUID,
    val vault_id: UUID,
    val name: String,
    val file_extension: String? = null,
    val mime_type: String? = null,
    val file_size: Long = 0,
    val storage_path: String? = null,
    val created_at: String,
    val uploaded_at: String,
    val last_modified_at: String? = null,
    val encryption_key_data: String? = null, // Base64 encoded
    val is_encrypted: Boolean = true,
    val document_type: String = "other",
    val source_sink_type: String? = null,
    val is_archived: Boolean = false,
    val is_redacted: Boolean = false,
    val status: String = "active",
    val extracted_text: String? = null,
    val ai_tags: List<String> = emptyList(),
    val file_hash: String? = null,
    val metadata: Map<String, String>? = null, // JSONB
    val author: String? = null,
    val camera_info: String? = null,
    val device_id: String? = null,
    val uploaded_by_user_id: UUID? = null,
    val updated_at: String
)

class SupabaseService {
    private var client: SupabaseClient? = null
    
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()
    
    private val _currentSession = MutableStateFlow<Session?>(null)
    val currentSession: StateFlow<Session?> = _currentSession.asStateFlow()
    
    private val _error = MutableStateFlow<Throwable?>(null)
    val error: StateFlow<Throwable?> = _error.asStateFlow()
    
    // Real-time channel subscriptions
    private val realtimeChannels = mutableMapOf<String, io.github.jan.supabase.realtime.channel.RealtimeChannel>()
    
    suspend fun configure() {
        try {
            client = createSupabaseClient(
                supabaseUrl = AppConfig.SUPABASE_URL,
                supabaseKey = AppConfig.SUPABASE_ANON_KEY
            ) {
                install(io.github.jan.supabase.postgrest.Postgrest)
                install(io.github.jan.supabase.realtime.Realtime)
                install(io.github.jan.supabase.storage.Storage)
                install(io.github.jan.supabase.auth.Auth)
            }
            
            // Test connection
            try {
                client?.from("users")
                    ?.select(columns = Columns.ALL) {
                        limit = 0
                    }
                
                _isConnected.value = true
                _currentSession.value = client?.auth?.currentSessionOrNull()
                Log.d("SupabaseService", "✅ Supabase client initialized and connected")
                Log.d("SupabaseService", "   URL: ${AppConfig.SUPABASE_URL}")
            } catch (e: Exception) {
                // Auth errors are expected when not signed in - connection is still OK
                _isConnected.value = true
                _currentSession.value = client?.auth?.currentSessionOrNull()
                Log.d("SupabaseService", "✅ Supabase client initialized")
                Log.d("SupabaseService", "   Note: No active session (user not signed in)")
            }
        } catch (e: Exception) {
            _isConnected.value = false
            _error.value = e
            Log.e("SupabaseService", "❌ Supabase initialization failed: ${e.message}")
        }
    }
    
    // MARK: - Authentication
    
    suspend fun signInWithGoogle(idToken: String): Session {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        
        return try {
            // Sign in with Google OAuth
            val session = client.auth.signInWith(Google) {
                this.idToken = idToken
            }
            
            _currentSession.value = session
            _isConnected.value = true
            
            if (AppConfig.ENABLE_REALTIME) {
                setupRealtimeSubscriptions()
            }
            
            Log.d("SupabaseService", "✅ Signed in with Google")
            session
        } catch (e: Exception) {
            Log.e("SupabaseService", "Sign in failed: ${e.message}")
            _error.value = e
            throw e
        }
    }
    
    suspend fun signOut() {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        
        try {
            client.auth.signOut()
            _currentSession.value = null
            _isConnected.value = false
            unsubscribeAll()
        } catch (e: Exception) {
            _error.value = e
            throw e
        }
    }
    
    suspend fun getCurrentUser(): UserInfo? {
        val client = client ?: return null
        return try {
            client.auth.currentUserOrNull()
        } catch (e: Exception) {
            Log.e("SupabaseService", "Error getting current user: ${e.message}")
            null
        }
    }
    
    fun currentSessionOrNull(): Session? {
        return _currentSession.value
    }
    
    // MARK: - Database Queries
    
    suspend inline fun <reified T> fetch(table: String, id: UUID): T {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        return client.from(table)
            .select {
                filter {
                    eq("id", id.toString())
                }
            }
            .decodeSingle<T>()
    }
    
    suspend inline fun <reified T> fetchAll(
        table: String,
        filters: Map<String, Any>? = null,
        limit: Int? = null,
        orderBy: String? = null,
        ascending: Boolean = true
    ): List<T> {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        
        // Build query with filters
        val query = if (filters != null && filters.isNotEmpty()) {
            client.from(table).select {
                filter {
                    filters.forEach { (key, value) ->
                        when (value) {
                            is String -> eq(key, value)
                            is UUID -> eq(key, value.toString())
                            is Int -> eq(key, value)
                            is Boolean -> eq(key, value)
                            else -> eq(key, value.toString())
                        }
                    }
                }
            }
        } else {
            client.from(table).select()
        }
        
        // Apply ordering
        val orderedQuery = if (orderBy != null) {
            query.order(orderBy, ascending = ascending)
        } else if (limit != null) {
            // Default order for limit
            query.order("created_at", ascending = false)
        } else {
            query
        }
        
        // Apply limit
        val finalQuery = if (limit != null) {
            orderedQuery.limit(limit)
        } else {
            orderedQuery
        }
        
        return finalQuery.decodeList<T>()
    }
    
    suspend inline fun <reified T> insert(table: String, values: T): T {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        return client.from(table)
            .insert(values) {
                select(Columns.ALL)
            }
            .decodeSingle<T>()
    }
    
    suspend inline fun <reified T> update(table: String, id: UUID, values: T): T {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        return client.from(table)
            .update(values) {
                filter {
                    eq("id", id.toString())
                }
                select(Columns.ALL)
            }
            .decodeSingle<T>()
    }
    
    suspend fun delete(table: String, id: UUID) {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        client.from(table)
            .delete {
                filter {
                    eq("id", id.toString())
                }
            }
    }
    
    // MARK: - Storage
    
    suspend fun uploadFile(bucket: String, path: String, data: ByteArray): String {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        client.storage.from(bucket).upload(path, data)
        return path
    }
    
    suspend fun downloadFile(bucket: String, path: String): ByteArray {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        return client.storage.from(bucket).downloadAuthenticated(path)
    }
    
    suspend fun deleteFile(bucket: String, path: String) {
        val client = client ?: throw IllegalStateException("Supabase client not initialized")
        client.storage.from(bucket).delete(path)
    }
    
    // MARK: - Real-time
    
    private fun setupRealtimeSubscriptions() {
        val client = client ?: return
        
        try {
            // Subscribe to vaults channel - specifically for threat_index updates
            val vaultsChannel = client.realtime.createChannel("vaults") {
                on("postgres_changes") { event ->
                    android.util.Log.d("SupabaseService", "Vault change received: ${event.type}")
                    
                    // Extract threat_index from update events
                    if (event.type == "UPDATE") {
                        try {
                            val payload = event.payload as? Map<*, *>
                            val newRecord = payload?.get("new") as? Map<*, *>
                            val threatIndex = newRecord?.get("threat_index") as? Number
                            val threatLevel = newRecord?.get("threat_level") as? String
                            val vaultId = newRecord?.get("id") as? String
                            
                            if (threatIndex != null && vaultId != null) {
                                android.util.Log.d("SupabaseService", "Threat index updated for vault $vaultId: $threatIndex ($threatLevel)")
                                // TODO: Emit to a Flow or callback for UI updates
                            }
                        } catch (e: Exception) {
                            android.util.Log.e("SupabaseService", "Error parsing vault update: ${e.message}")
                        }
                    }
                }
            }
            // Subscribe to vaults table changes, filtering for threat_index column updates
            vaultsChannel.subscribe()
            realtimeChannels["vaults"] = vaultsChannel
            
            // Subscribe to documents channel
            val documentsChannel = client.realtime.createChannel("documents") {
                on("postgres_changes") { event ->
                    android.util.Log.d("SupabaseService", "Document change received: ${event.type}")
                    // Handle document changes (INSERT, UPDATE, DELETE)
                }
            }
            documentsChannel.subscribe()
            realtimeChannels["documents"] = documentsChannel
            
            // Subscribe to nominees channel
            val nomineesChannel = client.realtime.createChannel("nominees") {
                on("postgres_changes") { event ->
                    android.util.Log.d("SupabaseService", "Nominee change received: ${event.type}")
                }
            }
            nomineesChannel.subscribe()
            realtimeChannels["nominees"] = nomineesChannel
            
            // Subscribe to chat_messages channel
            val chatChannel = client.realtime.createChannel("chat_messages") {
                on("postgres_changes") { event ->
                    android.util.Log.d("SupabaseService", "Chat message received: ${event.type}")
                }
            }
            chatChannel.subscribe()
            realtimeChannels["chat_messages"] = chatChannel
            
            // Subscribe to vault_sessions channel
            val sessionsChannel = client.realtime.createChannel("vault_sessions") {
                on("postgres_changes") { event ->
                    android.util.Log.d("SupabaseService", "Vault session change received: ${event.type}")
                }
            }
            sessionsChannel.subscribe()
            realtimeChannels["vault_sessions"] = sessionsChannel
            
            android.util.Log.d("SupabaseService", "✅ Real-time subscriptions established for ${realtimeChannels.size} channels")
        } catch (e: Exception) {
            android.util.Log.e("SupabaseService", "❌ Failed to set up real-time subscriptions: ${e.message}")
        }
    }
    
    private fun unsubscribeAll() {
        try {
            realtimeChannels.values.forEach { channel ->
                try {
                    channel.unsubscribe()
                } catch (e: Exception) {
                    android.util.Log.e("SupabaseService", "Error unsubscribing from channel: ${e.message}")
                }
            }
            realtimeChannels.clear()
            android.util.Log.d("SupabaseService", "✅ Unsubscribed from all real-time channels")
        } catch (e: Exception) {
            android.util.Log.e("SupabaseService", "❌ Error unsubscribing from channels: ${e.message}")
        }
    }
}

